hnix-store-remote/src/System/Nix/Store/Remote/Parsers.hs

```diff
@@ -2,63 +2,54 @@
 {-# LANGUAGE OverloadedStrings   #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE DataKinds           #-}
 {-# LANGUAGE TypeApplications    #-}
 
-module System.Nix.Store.Remote.Parsers (
-    parseContentAddressableAddress
-  ) where
+module System.Nix.Store.Remote.Parsers
+  ( parseContentAddressableAddress
+  )
+where
 
-import Control.Applicative ((<|>))
-import Data.Attoparsec.ByteString.Char8 (Parser, (<?>))
-import Data.ByteString (ByteString)
-import System.Nix.Hash (Digest, NamedAlgo, SomeNamedDigest(SomeDigest))
-import System.Nix.StorePath (ContentAddressableAddress(..), NarHashMode(..))
-
-import qualified Data.Attoparsec.ByteString.Char8
-import qualified Data.ByteString.Char8
+import           Control.Applicative            ( (<|>) )
+import           Data.Attoparsec.ByteString.Char8
+import           Data.ByteString.Char8
 import qualified Data.Text
-
-import qualified System.Nix.Internal.Base32
-import qualified System.Nix.Hash
+import           Data.Text.Encoding             ( decodeUtf8 )
+import           System.Nix.Hash
+import           System.Nix.StorePath           ( ContentAddressableAddress(..)
+                                                , NarHashMode(..)
+                                                )
 
 -- | Parse `ContentAddressableAddress` from `ByteString`
-parseContentAddressableAddress :: forall hashAlgo . NamedAlgo hashAlgo
-                              => ByteString
-                              -> Either String ContentAddressableAddress
+parseContentAddressableAddress
+  :: ByteString -> Either String ContentAddressableAddress
 parseContentAddressableAddress =
-  Data.Attoparsec.ByteString.Char8.parseOnly
-    (contentAddressableAddressParser @hashAlgo)
+  Data.Attoparsec.ByteString.Char8.parseOnly contentAddressableAddressParser
 
 -- | Parser for content addressable field
-contentAddressableAddressParser :: forall hashAlgo . NamedAlgo hashAlgo
-                                => Parser ContentAddressableAddress
-contentAddressableAddressParser =
-      caText
-  <|> caFixed @hashAlgo
+contentAddressableAddressParser :: Parser ContentAddressableAddress
+contentAddressableAddressParser = caText <|> caFixed
 
 -- | Parser for @text:sha256:<h>@
 caText :: Parser ContentAddressableAddress
 caText = do
-  _ <- "text:sha256:"
-  digest <- parseDigest
-  either fail return
-    $ Text <$> digest
+  _      <- "text:sha256:"
+  digest <- decodeBase32 @'SHA256 <$> parseHash
+  either fail return $ Text <$> digest
 
 -- | Parser for @fixed:<r?>:<ht>:<h>@
-caFixed :: forall hashAlgo . NamedAlgo hashAlgo => Parser ContentAddressableAddress
+caFixed :: Parser ContentAddressableAddress
 caFixed = do
-  _ <- "fixed:"
-  narHashMode <- (pure Recursive <$> "true") <|> (pure RegularFile <$> "false")
-    <?> "Invalid Base32 part"
-  digest <- parseDigest
+  _           <- "fixed:"
+  narHashMode <- (pure Recursive <$> "r:") <|> (pure RegularFile <$> "")
```

Comment:
Not too sure here. Maybe it's really "true" and "false" ? In mkFixedOutputDerivation we also have a name after fixed:, as in fixed:out:...

How can I properly test this ?

Reply:
Looks good, in my repo I have a similar fix pending:
```
narHashMode <- Data.Attoparsec.ByteString.Char8.option RegularFile (pure Recursive <$> "r:")
```
but I've stopped at the `digest` since it required changes from this PR.

For testing this I'll draft `hnix-store-db` PR shortly.